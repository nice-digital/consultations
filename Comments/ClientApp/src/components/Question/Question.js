// @flow

import React, { Component } from "react";
import ReactMarkdown from "react-markdown";
import { Answer } from "../Answer/Answer";
import { UserContext } from "../../context/UserContext";

type PropsType = {
	staticContext?: any,
	saveAnswerHandler: Function,
	deleteAnswerHandler: Function,
	question: QuestionType,
	otherUsersAnswers: Array<AnswerType>,
	readOnly: boolean,
	isUnsaved: boolean,
	documentTitle?: string,
	showAnswer: boolean,
};
export class Question extends Component<PropsType> {

	isTextSelection = (question) => question.commentOn && question.commentOn.toLowerCase() === "selection" && question.quote;

	returnAnswers = (answers, readOnly = this.props.readOnly) => {
		const {
			question,
			updateUnsavedIds,
			saveAnswerHandler,
			deleteAnswerHandler,
		} = this.props;

		const answerComponents = answers.map(answer => (
			<Answer
				questionText={question.questionText}
				questionType={question.questionType}
				updateUnsavedIds={updateUnsavedIds}
				questionId={question.questionId}
				readOnly={readOnly}
				key={answer.answerId}
				unique={`Question${question.questionId}-Answer${answer.answerId}`}
				answer={answer}
				saveAnswerHandler={saveAnswerHandler}
				deleteAnswerHandler={deleteAnswerHandler}
			/>
		));

		return answerComponents;
	};

	render() {
		if (!this.props.question) return null;

		const { documentTitle } = this.props;
		const { commentOn, quote } = this.props.question;

		let otherUsersAnswers = this.props.otherUsersAnswers || [];
		const answers = this.props.question.answers || [];
		let answersToShow = answers.filter(answer => answer.showWhenFiltered);
		if (answersToShow === null || answersToShow.length < 1){
			answersToShow = [{
				answerId: -1,
				questionId: this.props.question.questionId,
				commenterEmail: null,
			}];
		}

		return (
			<li className={this.props.isUnsaved ? "CommentBox CommentBox--unsavedChanges" : "CommentBox"}>
				<>
					{documentTitle &&
						<h3 className="CommentBox__title mt--0 mb--0">{documentTitle}</h3>
					}
					<h4 data-qa-sel="comment-box-title" className="CommentBox__title mt--0 mb--0">
						Question on <span className="text-lowercase">{commentOn}</span>
					</h4>
				</>

				{this.isTextSelection(this.props.question) &&
					<>
						<h3 className="CommentBox__title mt--0 mb--0">{documentTitle}</h3>
						<h4 data-qa-sel="comment-box-title" className="CommentBox__title mt--0 mb--0">
							Question on: <span className="text-lowercase">{commentOn}</span>
						</h4>
						<div className="CommentBox__quote mb--d">{quote}</div>
					</>
				}

				<div className="font-weight-bold markdown mt--d mb--d">
					<ReactMarkdown
						allowedTypes={["root", "text", "list", "listItem"]}
						unwrapDisallowed={true}
						source={this.props.question.questionText}
					/>
				</div>

				{this.props.isUnsaved &&
					<p className="CommentBox__validationMessage" role="alert">You have unsaved changes</p>
				}
				<UserContext.Consumer>
					{(contextValue: ContextType) => {
						return (
							this.props.showAnswer && (
								<>
									{this.returnAnswers(answersToShow)}
									{!contextValue.isLead &&
										this.returnAnswers(otherUsersAnswers, true)
									}
								</>
							)
						);}}
				</UserContext.Consumer>

			</li>
		);
	}
}
