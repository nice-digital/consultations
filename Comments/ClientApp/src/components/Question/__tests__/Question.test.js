import React from "react";
import { render } from "@testing-library/react";
import { Question } from "../Question";
import questionWithAnswer from "./questionWithAnswer.json";
import questionWithoutAnswer from "./questionWithoutAnswer.json";

test("should match snapshot with answer", () => {
	const fakePropsWithAnswer = {
		readOnly: false,
		key: 1,
		unique: "unique",
		question: questionWithAnswer,
		showAnswer: true,
	};
	const {container} = render(<Question {...fakePropsWithAnswer} />);
	expect(container).toMatchSnapshot();
});

test("should match snapshot without answer", () => {
	const fakePropsWithoutAnswer = {
		readOnly: false,
		key: 1,
		unique: "unique",
		question: questionWithoutAnswer,
		showAnswer: false,
	};
	const {container} = render(<Question {...fakePropsWithoutAnswer} />);
	expect(container).toMatchSnapshot();
});
