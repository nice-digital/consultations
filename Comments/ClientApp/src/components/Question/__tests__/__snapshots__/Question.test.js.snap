// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with answer 1`] = `
<div>
  <li
    class="CommentBox"
  >
    <h4
      class="CommentBox__title mt--0 mb--0"
      data-qa-sel="comment-box-title"
    >
      Question on 
      <span
        class="text-lowercase"
      >
        Consultation
      </span>
    </h4>
    <div
      class="font-weight-bold markdown mt--d mb--d"
    >
      Which areas will have the biggest impact on practice and be challenging to implement? Please say for whom and why.
    </div>
    <section
      role="form"
    >
      <form
        class="mb--0 mt--e"
      />
    </section>
  </li>
</div>
`;

exports[`should match snapshot without answer 1`] = `
<div>
  <li
    class="CommentBox"
  >
    <h4
      class="CommentBox__title mt--0 mb--0"
      data-qa-sel="comment-box-title"
    >
      Question on 
      <span
        class="text-lowercase"
      >
        Consultation
      </span>
    </h4>
    <div
      class="font-weight-bold markdown mt--d mb--d"
    >
      Which areas will have the biggest impact on practice and be challenging to implement? Please say for whom and why.
    </div>
  </li>
</div>
`;
